From 16832f7c1baf0f446e77a98f578b6452e96b94e9 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <grinberg@compulab.co.il>
Date: Tue, 26 Feb 2013 18:40:10 +0200
Subject: ARM: i.MX6: cm-fx6: reset the on-board USB hub

cm-fx6 features an on-board USB hub.
Reset the hub as a part of boot process.

Signed-off-by: Igor Grinberg <grinberg@compulab.co.il>
---
 arch/arm/mach-mx6/board-cm-fx6-dl.h |    2 ++
 arch/arm/mach-mx6/board-cm-fx6-q.h  |    2 ++
 arch/arm/mach-mx6/board-cm-fx6.c    |   19 +++++++++++++++++++
 3 files changed, 23 insertions(+)

diff --git a/arch/arm/mach-mx6/board-cm-fx6-dl.h b/arch/arm/mach-mx6/board-cm-fx6-dl.h
index ce7d69b..a6f7d3c 100644
--- a/arch/arm/mach-mx6/board-cm-fx6-dl.h
+++ b/arch/arm/mach-mx6/board-cm-fx6-dl.h
@@ -26,6 +26,8 @@ static iomux_v3_cfg_t cm_fx6_dl_pads[] = {
 	MX6DL_PAD_EIM_EB3__GPIO_2_31,
 	/* TSPENDOWN */
 	MX6DL_PAD_SD4_DAT7__GPIO_2_15,
+	/* USB HUB Reset */
+	MX6DL_PAD_SD3_RST__GPIO_7_8,
 	/* USBH1 VBUS control */
 	MX6DL_PAD_GPIO_0__USBOH3_USBH1_PWR,
 
diff --git a/arch/arm/mach-mx6/board-cm-fx6-q.h b/arch/arm/mach-mx6/board-cm-fx6-q.h
index ceb3931..eb2fb54 100644
--- a/arch/arm/mach-mx6/board-cm-fx6-q.h
+++ b/arch/arm/mach-mx6/board-cm-fx6-q.h
@@ -26,6 +26,8 @@ static iomux_v3_cfg_t cm_fx6_q_common_pads[] = {
 	MX6Q_PAD_EIM_EB3__GPIO_2_31,
 	/* TSPENDOWN */
 	MX6Q_PAD_SD4_DAT7__GPIO_2_15,
+	/* USB HUB Reset */
+	MX6Q_PAD_SD3_RST__GPIO_7_8,
 	/* USBH1 VBUS control */
 	MX6Q_PAD_GPIO_0__USBOH3_USBH1_PWR,
 
diff --git a/arch/arm/mach-mx6/board-cm-fx6.c b/arch/arm/mach-mx6/board-cm-fx6.c
index 2962020..d53dd3c 100644
--- a/arch/arm/mach-mx6/board-cm-fx6.c
+++ b/arch/arm/mach-mx6/board-cm-fx6.c
@@ -83,6 +83,7 @@
 #define CM_FX6_ECSPI1_CS0		IMX_GPIO_NR(2, 30)
 #define CM_FX6_GREEN_LED		IMX_GPIO_NR(2, 31)
 #define CM_FX6_ECSPI1_CS1		IMX_GPIO_NR(3, 19)
+#define CM_FX6_USB_HUB_RST		IMX_GPIO_NR(7, 8)
 
 #define MX6_ARM2_LDB_BACKLIGHT		IMX_GPIO_NR(1, 9)
 #define MX6_ARM2_USB_OTG_PWR		IMX_GPIO_NR(3, 22)
@@ -426,6 +427,22 @@ static void imx6_arm2_usbotg_vbus(bool on)
 		gpio_set_value(MX6_ARM2_USB_OTG_PWR, 0);
 }
 
+static void __init cm_fx6_usb_hub_reset(void)
+{
+	int err;
+
+	err = gpio_request_one(CM_FX6_USB_HUB_RST,
+			       GPIOF_OUT_INIT_LOW, "usb hub rst");
+	if (err) {
+		pr_err("CM-FX6: usb hub rst gpio request failed: %d\n", err);
+		return;
+	}
+
+	udelay(10);
+	gpio_set_value(CM_FX6_USB_HUB_RST, 1);
+	msleep(1);
+}
+
 static void __init cm_fx6_init_usb(void)
 {
 	int ret = 0;
@@ -441,6 +458,8 @@ static void __init cm_fx6_init_usb(void)
 		pr_err("%s: USB_OTG_PWR gpio request failed: %d\n",
 		       __func__, ret);
 
+	cm_fx6_usb_hub_reset();
+
 	mxc_iomux_set_gpr_register(1, 13, 1, 1);
 
 	mx6_set_otghost_vbus_func(imx6_arm2_usbotg_vbus);
-- 
1.7.9.5

